design.md

```diff
@@ -1,13 +1,79 @@
 # Design
 
+A grammar for our language:
+```
+<program>    := <state> <program>
+              | <state>
+
+<state>      := state <state_name> <rules>
+
+<rules>      := <rule> <rules>
+              | <rule>
+
+<rule>       := "|" move <dir>
+              | "|" face <dir>
+              | "|" <state_name>
+              | "|" move <dir>, face <dir>
+              | "|" move <dir>, <state_name>
+              | "|" face <dir>, <state_name>
+              | "|" move <dir>, face <dir> <state_name>
+
+<dir>        := north | south | east | west | left | right | forward | backward
+
+<state_name> := [a-zA-Z0-9_]+
+```
```

Comment:
Ideally you should know the grammar of your language (at least informally) before you start trying to parse it.